### PR TITLE
[feat] Adding context propagation to net/http server instrumentation

### PR DESCRIPTION
**Pull request by oxeye-gal**
Hi,

This PR attempts to add context propagation to the net/http server instrumentation.

We do it by iterating over the buckets of the incoming request header map and trying to find the "traceparent" header.

This was tested with Go versions 1.12-1.19.

Our test environment consisted of instrumented Python application that sends a request to instrumented Go http server.

The produced trace looks like the following:

213373358-c184e90f-597c-4ca0-ba3f-52365fc89770

We would love to get your feedback on this.
Thanks!

### CODE REVIEW

**Review by damemi — DISMISSED** [review]
This looks like you need to run `make generate` based on the [failure](https://github.com/open-telemetry/opentelemetry-go-instrumentation/actions/runs/4783444662/jobs/8520732823?pr=92#step:6:58)

Also the e2e failure is interesting, because it seems that this change removed the redundant empty span that @JamieDanielson was trying to fix in https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/86 (see [test output](https://github.com/open-telemetry/opentelemetry-go-instrumentation/actions/runs/4783444659/jobs/8520732629?pr=92#step:17:33))

**Comment by damemi** [comment]
> also the redundant empty span removed is due to the removal of the net/http.(*ServeMux).ServeHTTP symbol from the functions to instrument as also mentioned by https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/86

Ah I see. Is that necessary for context propagation to work in this change?

**Comment by JamieDanielson** [comment]
> > also the redundant empty span removed is due to the removal of the net/http.(*ServeMux).ServeHTTP symbol from the functions to instrument as also mentioned by #86
> 
> Ah I see. Is that necessary for context propagation to work in this change?

Is it cleaner to keep both function names in this PR so the output of the test remains the same, and I can update the other PR after this is merged in? That way we separate concerns, because I don't _think_ the extra empty span has an impact on this one (though I could be mistaken, I've been working on a few different branches of branches lately).

**Comment by damemi** [comment]
> Is it cleaner to keep both function names in this PR so the output of the test remains the same, and I can update the other PR after this is merged in? That way we separate concerns

Yeah that's what I was wondering. if it's relevant to this PR, we can keep it, but otherwise @oxeye-gal can you remove that change so we don't duplicate the work?

**Comment by MikeGoldsmith** [comment]
Hey @oxeye-gal - this PR needs rebasing after we merged ARM support on #82. I’m happy to help do that if you would like us to.

**Review by MikeGoldsmith — APPROVED** [review]
Looks great, thanks for the contribution @oxeye-gal. I'll look at updating #91 next 👍🏻

**Comment by oxeye-gal** [comment]
@MikeGoldsmith seems like the build fails again due to the added Request.Header offset, what is the correct way to do it? thanks again for the help!

**Comment by MikeGoldsmith** [comment]
After you've updated the offsets, you'll need to update the offsets.json. This can be done using `make docker-offsets`.

**Comment by oxeye-gal** [comment]
I have added the new field to offsets-tracker, I hope this fixes it @MikeGoldsmith

**Comment by MikeGoldsmith** [comment]
We're having hard time with this one 👎🏻  I'll give it a go too

File: include/go_types.h
Comment (pdelewski) [inline include/go_types.h:51]: Is it used somewhere?

Comment (pdelewski) [inline include/go_types.h:51]: Is it used somewhere?

Comment (oxeye-gal) [reply include/go_types.h:51]: You mean the whole map_bucket struct or specifically the overflow field?

The map_bucket struct is used by the context propagation code, and even if the overflow field is not directly used by our code it is still important to keep it as part of the map_bucket struct as we are iterating through the buckets of the map, and if remove this field we will end up reading wrong values from memory.

Comment (pdelewski) [reply include/go_types.h:51]: I meant `overflow` field. Could you clarify why 

> we will end up reading wrong values from memory.

?

File: pkg/instrumentors/bpf/net/http/server/bpf/probe.bpf.c
Comment (pdelewski) [inline pkg/instrumentors/bpf/net/http/server/bpf/probe.bpf.c:114]: Would not make sense to extract `8` as constant? and use it also in struct

File: include/go_types.h
Comment (oxeye-gal) [reply include/go_types.h:51]: Sure, I used the following resource to understand the Go map implementation:

https://github.com/cch123/golang-notes/blob/master/map.md

As you can see the map bucket struct has an overflow field, if we omit it from the map_bucket struct line 110 in the probe.bpf.c file which attempts to read the next bucket will not read it correctly as the loop advances.

**Comment by MikeGoldsmith** [comment]
@open-telemetry/go-instrumentation-approvers this is ready for another review 👍🏻

File: pkg/instrumentors/bpf/net/http/server/bpf/probe.bpf.c
Comment (MikeGoldsmith) [reply pkg/instrumentors/bpf/net/http/server/bpf/probe.bpf.c]: @pdelewski I added a specific TODO when consolidating the HTTP probes to ensure const values are used for max loop iterations. Resolving this comment so we can get this merged then refactor.

**Review by edeNFed — APPROVED** [review]
lgtm

**Review by MrAlias — APPROVED** [review]

**Pull request merged**
